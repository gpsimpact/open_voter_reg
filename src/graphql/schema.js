import glob from 'glob';
import fs from 'fs';
import path from 'path';
import { makeExecutableSchema } from 'graphql-tools';

process.on('warning', e => console.warn(e.stack));

const typeDefs = [];

// loop over types and import the contents
glob.sync('**/*.type.gql', { cwd: __dirname }).forEach(filename => {
  const filePath = path.join(__dirname, filename);
  const fileContent = fs.readFileSync(filePath, { encoding: 'utf8' });
  typeDefs.push(fileContent);
});

import resolvers from './resolvers';

// create a schema
const schema = makeExecutableSchema({ typeDefs, resolvers });

// add mocks to the schema, preserving the existing resolvers (none for the time beeing)
// import mocks from './mocks';
// addMockFunctionsToSchema({ schema, mocks, preserveResolvers: true });

export default schema;
