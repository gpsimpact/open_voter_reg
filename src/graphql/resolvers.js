import QueryResolver from './Query/Query.resolvers';
import MutationResolver from './Mutation/Mutation.resolvers';

export default {
  ...QueryResolver,
  ...MutationResolver,
};
