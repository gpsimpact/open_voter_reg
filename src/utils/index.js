import faker from 'faker';
import { includes } from 'lodash';
import { InsufficientPermissionsError } from '../errors';

export function mapTo(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), row));
    return Array.from(group.values());
  };
}

export function mapToMany(keys, keyFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, []]));
    rows.forEach(row => (group.get(keyFn(row)) || []).push(row));
    return Array.from(group.values());
  };
}

export function mapToValues(keys, keyFn, valueFn) {
  return rows => {
    const group = new Map(keys.map(key => [key, null]));
    rows.forEach(row => group.set(keyFn(row), valueFn(row)));
    return Array.from(group.values());
  };
}

export const generateFakeUsers = (num, seed) => {
  faker.seed(seed); // ensures consistent result
  const fakeUsers = [];
  Array(num)
    .fill()
    .map((_, i) => {
      fakeUsers.push({
        first_name: faker.name.firstName(),
        last_name: faker.name.lastName(),
        email: faker.internet.email(),
        password: faker.internet.password(),
      });
    });
  return fakeUsers;
};

export const hasPermission = (viewer, service, requiredPermission, throwOnFail = false) => {
  if (!includes(viewer.permissions[service], requiredPermission)) {
    if (throwOnFail) {
      throw new InsufficientPermissionsError();
    }
    return false;
  }
  return true;
};
