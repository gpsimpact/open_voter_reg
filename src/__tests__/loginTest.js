import { graphql } from 'graphql';
import db from '../db';
import schema from '../graphql/schema';
import faker from 'faker';
import MakeContext from '../Context';
import { find } from 'lodash';
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { generateFakeUsers } from '../utils';

beforeAll(async () => await db.migrate.latest({ directory: './src/db/migrations' }));
beforeEach(
  async () =>
    await Promise.all([
      db.raw('TRUNCATE TABLE permissions CASCADE'),
      db.raw('TRUNCATE TABLE organizations CASCADE'),
      db.raw('TRUNCATE TABLE users CASCADE'),
    ])
);
afterAll(async () => await db.destroy());

describe('login', () => {
  test('user can login via mutation', async () => {
    const users = generateFakeUsers(1, 1);
    // generate old password hash
    const passwordHash = bcrypt.hashSync(users[0].password, 10);
    await db('users').insert({ email: users[0].email, password: passwordHash });
    const org1 = faker.random.uuid();
    const org2 = faker.random.uuid();
    await db('organizations').insert({ hash: org1, name: faker.company.companyName() });
    await db('organizations').insert({ hash: org2, name: faker.company.companyName() });
    await db('permissions').insert({
      org_hash: org1,
      email: users[0].email,
      permission: 'READ_ONLY',
    });
    await db('permissions').insert({
      org_hash: org2,
      email: users[0].email,
      permission: 'ADMIN',
    });
    const tokenPayload = {
      email: users[0].email,
      permissions: {
        [org1]: ['READ_ONLY'],
        [org2]: ['ADMIN'],
      },
    };
    const query = `
        mutation {
            login(
              email: "${users[0].email}",
              password: "${users[0].password}"
            ) {
              userProfile {
                email
              }
              token
            }
          }
      `;
    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    // console.log(JSON.stringify(result, null, '\t'));
    const decodedToken = jwt.verify(result.data.login.token, process.env.JWT_SECRET);
    expect(decodedToken.permissions).toEqual(tokenPayload.permissions);
    expect(decodedToken.email).toEqual(tokenPayload.email);
  });

  test('user can not log in with incorrect password', async () => {
    const users = generateFakeUsers(1, 21);
    // generate old password hash
    const passwordHash = bcrypt.hashSync(users[0].password, 10);
    await db('users').insert({ email: users[0].email, password: passwordHash });
    const query = `
        mutation {
            login(
              email: "${users[0].email}",
              password: "badpass"
            ) {
              userProfile {
                email
              }
              token
            }
          }
      `;
    const rootValue = {};
    const context = new MakeContext({ user: null });
    const result = await graphql(schema, query, rootValue, context);
    expect(result).toMatchSnapshot();
    expect(result.errors.length).toBeGreaterThanOrEqual(1);
    expect(find(result.errors, { message: 'User can not be authenticated.' })).not.toBeUndefined();
  });
});
