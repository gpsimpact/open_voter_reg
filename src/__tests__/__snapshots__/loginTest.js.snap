// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login user can not log in with incorrect password 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    [GraphQLError: User can not be authenticated.],
  ],
}
`;
